src/librustc/mir/cache.rs

```diff
@@ -29,43 +32,335 @@ impl<'a> HashStable<StableHashingContext<'a>> for Cache {
+                    }
+                }
+            }
+
+            self.predecessors = Some(result)
         }
+    }
+
+    #[inline]
+    /// This will recompute the predecessors cache if it is not available
+    fn predecessors(&mut self, body: &Body<'_>) -> &IndexVec<BasicBlock, Vec<BasicBlock>> {
+        self.ensure_predecessors(body);
+        self.predecessors.as_ref().unwrap()
+    }
 
-        ReadGuard::map(self.predecessors.borrow(), |p| p.as_ref().unwrap())
+    #[inline]
+    fn predecessors_for(&mut self, bb: BasicBlock, body: &Body<'_>) -> &[BasicBlock] {
+        &self.predecessors(body)[bb]
+    }
+
+    #[inline]
+    fn unwrap_predecessors_for(&self, bb: BasicBlock) -> &[BasicBlock] {
+        &self.predecessors.as_ref().unwrap()[bb]
+    }
+
+    #[inline]
+    impl_predecessor_locations!((pub) predecessor_locations mut);
+
+    impl_predecessor_locations!(() unwrap_predecessor_locations);
+
+    #[inline]
+    pub fn basic_blocks_mut<'a, 'tcx>(
+        &mut self,
+        body: &'a mut Body<'tcx>
+    ) -> &'a mut IndexVec<BasicBlock, BasicBlockData<'tcx>> {
+        debug!("bbm: Clearing predecessors cache for body at: {:?}", body.span.data());
+        self.invalidate_predecessors();
+        &mut body.basic_blocks
+    }
+
+    #[inline]
+    pub fn basic_blocks_and_local_decls_mut<'a, 'tcx>(
+        &mut self,
+        body: &'a mut Body<'tcx>
+    ) -> (&'a mut IndexVec<BasicBlock, BasicBlockData<'tcx>>, &'a mut LocalDecls<'tcx>) {
+        debug!("bbaldm: Clearing predecessors cache for body at: {:?}", body.span.data());
+        self.invalidate_predecessors();
+        (&mut body.basic_blocks, &mut body.local_decls)
     }
 }
 
-fn calculate_predecessors(body: &Body<'_>) -> IndexVec<BasicBlock, Vec<BasicBlock>> {
-    let mut result = IndexVec::from_elem(vec![], body.basic_blocks());
-    for (bb, data) in body.basic_blocks().iter_enumerated() {
-        if let Some(ref term) = data.terminator {
-            for &tgt in term.successors() {
-                result[tgt].push(bb);
-            }
+#[derive(Clone, Debug, RustcEncodable, RustcDecodable)]
+pub struct BodyCache<'tcx> {
+    cache: Cache,
+    body: Body<'tcx>,
+}
+
+impl BodyCache<'tcx> {
+    pub fn new(body: Body<'tcx>) -> Self {
+        Self {
+            cache: Cache::new(),
+            body,
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! read_only {
+    ($body_cache:expr) => {
+        {
+            $body_cache.ensure_predecessors();
+            $body_cache.unwrap_read_only()
         }
+    };
+}
+
+impl BodyCache<'tcx> {
+    pub fn ensure_predecessors(&mut self) {
+        self.cache.ensure_predecessors(&self.body);
+    }
+
+    pub fn predecessors(&mut self) -> &IndexVec<BasicBlock, Vec<BasicBlock>> {
+        self.cache.predecessors(&self.body)
+    }
+
+    pub fn unwrap_read_only(&self) -> ReadOnlyBodyCache<'_, 'tcx> {
+        ReadOnlyBodyCache::new(&self.cache, &self.body)
+    }
+
+    pub fn body(&self) -> &Body<'tcx> {
```

Comment:
I don't think we should have this function, we should rely on the `Deref` impl instead

Reply:
Ok, that'll require me to impl the traits. I'll get started on that.

Reply:
I could only remove this impl on `BodyCache`. The one on `ReadOnlyBodyCache` couldn't be removed because the lifetime of the returned refrence in `Deref` is not the same.

Reply:
oh right :( that's a bit unfortunate. Having this mix between using deref and using `.body()` seems like it could cause confusion, but at the same time having the `Deref` impl is such a boon. I guess you can just keep what you have.

> Ok, that'll require me to impl the traits. I'll get started on that.

Oh no, that's not necessary. If you have a `&BodyCache`, you can use `&**foo` to get at the `&Body`